ddtrace/bootstrap/sitecustomize.py

```diff
@@ -2,30 +2,112 @@
 Bootstrapping code that is run when using the `ddtrace-run` Python entrypoint
 Add all monkey-patching that needs to run by default here
 """
-import logging
-import os
 import sys
-from typing import Any
-from typing import Dict
 
 
-# Perform gevent patching as early as possible in the application before
-# importing more of the library internals.
-if os.environ.get("DD_GEVENT_PATCH_ALL", "false").lower() in ("true", "1"):
-    import gevent.monkey
+MODULES_LOADED_AT_STARTUP = frozenset(sys.modules.keys())
+MODULES_THAT_TRIGGER_CLEANUP_WHEN_INSTALLED = ("gevent",)
 
-    gevent.monkey.patch_all()
 
+import os  # noqa
+
+
+MODULES_TO_NOT_CLEANUP = {"atexit", "asyncio", "attr", "concurrent", "ddtrace", "logging"}
+if sys.version_info <= (2, 7):
+    MODULES_TO_NOT_CLEANUP |= {"encodings", "codecs"}
+    import imp
+
+    _unloaded_modules = []
+else:
+    import importlib
+
+
+def is_installed(module_name):
+    # https://stackoverflow.com/a/51491863/735204
+    if sys.version_info >= (3, 4):
+        return importlib.util.find_spec(module_name)
+    elif sys.version_info >= (3, 3):
+        return importlib.find_loader(module_name)
+    elif sys.version_info >= (3, 1):
+        return importlib.find_module(module_name)
+    elif sys.version_info >= (2, 7):
+        try:
+            imp.find_module(module_name)
+        except ImportError:
+            return False
+        else:
+            return True
+    return False
+
+
+def should_cleanup_loaded_modules():
+    dd_unload_sitecustomize_modules = os.getenv("DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE", default="0").lower()
+    if dd_unload_sitecustomize_modules == "0":
+        return False
+    elif dd_unload_sitecustomize_modules not in ("1", "auto"):
+        return False
+    elif dd_unload_sitecustomize_modules == "auto" and not any(
+        is_installed(module_name) for module_name in MODULES_THAT_TRIGGER_CLEANUP_WHEN_INSTALLED
+    ):
+        return False
+    return True
+
+
+def cleanup_loaded_modules(aggressive=False):
+    """
+    "Aggressive" here means "cleanup absolutely every module that has been loaded since startup".
+    Non-aggressive cleanup entails leaving untouched certain modules
+    This distinction is necessary because this function is used both to prepare for gevent monkeypatching
+    (requiring aggressive cleanup) and to implement "module cloning" (requiring non-aggressive cleanup)
+    """
+    modules_loaded_since_startup = set(_ for _ in sys.modules if _ not in MODULES_LOADED_AT_STARTUP)
+    # Figuring out modules_loaded_since_startup is necessary because sys.modules has more in it than just what's in
+    # import statements in this file, and unloading some of them can break the interpreter.
```

Comment:
> unloading some of them can break the interpreter

this statement makes me feeling like `aggressive=True` is dangerous and should not be used. maybe I don't fully understand the difference and why we need aggressive for pre-patching gevent.

Reply:
I don't understand the specific reason that aggressive unloading is necessary for gevent patching. The best I can offer is that it aligns with gevent's [guidance](https://www.gevent.org/api/gevent.monkey.html)

> Warning Patching too late can lead to unreliable behaviour (for example, some modules may still use blocking sockets) or even errors.

Tests [fail in confusing ways](https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/28692/workflows/99c542b0-0fbe-4106-bc60-a839805e9afc/jobs/1951470) when aggressive unloading isn't used.

Aggressive unloading isn't particularly dangerous because it does not unload modules that were loaded at startup. That's the thing that can break the interpreter.